### PR TITLE
Serialize queries and export their types and schema

### PR DESCRIPTION
**Pull request by arielshaqed**

### CODE REVIEW

File: db/src/query.ts [hunk 1]
```diff
@@ -1,89 +1,15 @@
...
-
-export type Filter = EqFilter | NeFilter
-  | GtFilter | GteFilter | LtFilter | LteFilter
-  | ExistsFilter | IsNullFilter
-  | MatchesFilter | StartsWithFilter
-  | AndFilter | OrFilter | NotFilter;
+// To avoid injection filters are marked during *construction*, but
+// that marking is not required for use.  The DB code is on the other
+// side of a serialization barrier.
+export type Marked<T extends {}> = T & { [filterSymbol]: true };
```
Comment (bergundy) [inline db/src/query.ts]: I'd redefine all the Marked versions of the filters here:
```typescript
export type EqFilter = Marked<dbi.EqFilter>;
...
export interface AndFilter extends Marked<dbi.AndFilter> {
  filters: Filter[];
}
...
export type Filter = EqFilter | ... | AndFilter;
```

Comment (bergundy) [reply db/src/query.ts]: Please also reexport Order, DESC and ASC.

File: interfaces/package.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "@binaris/shift-interfaces",
+  "version": "0.0.1",
+  "files": [
+    "dist"
+  ],
+  "description": "interfaces for serving ShiftJS",
+  "main": "dist/index.js",
+  "types": "dist/index.d.js",
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply interfaces/package.json]: Bad copy-paste, thanks!

Sending mini-PR to bad copy source.

File: interfaces/package.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  "scripts": {
+    "build": "rm -rf dist/ && tsc && npm run compile-schema",
+    "compile-schema": "typescript-json-schema -o dist/schema.json --required --include src/db.ts --noExtraProps tsconfig.json '*'",
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply interfaces/package.json]: Did the first, ack the second.

**Comment by arielshaqed** [comment]
@bergundy PTAL

File: db/src/query.ts [hunk 4]
```diff
@@ -234,14 +191,14 @@ class Path {
    * @param caseInsensitive should the check be case insensitive?
    * @return - a filter
    */
-  public matches(pattern: string, caseInsensitive?: boolean): MatchesFilter;
+  public matches(pattern: string, caseInsensitive?: boolean): dbi.MatchesFilter;
```
Comment (bergundy) [inline db/src/query.ts]: You forgot this one..

File: db/src/query.ts [hunk 5]
```diff
@@ -125,7 +82,7 @@ class Path {
    * @param x - value for comparison
    * @return - a filter
    */
-  public eq(x: Equatable): EqFilter {
+  public eq(x: dbi.Equatable): EqFilter {
```
Comment (bergundy) [inline db/src/query.ts]: I'd import Equatable and Comparable

File: db/package.json [hunk 6]
```diff
@@ -15,6 +15,7 @@
   "author": "",
   "license": "MIT",
   "dependencies": {
+    "@binaris/shift-interfaces": "0.0.1",
```
[the remark this answers is not in the document]
Comment (vogre) [reply db/package.json]: This intentionally the exact version. When interfaces version changes during publication also the db version changes, this allows us to know exactly what are the versions of dependant packages from the version of the parent instead of a combinatorial guess.

Note that even if interfaces change is major, the automatic change in db package will be patch.

File: db/src/query.ts (hunk 5 above)
Comment (arielshaqed) [reply db/src/query.ts]: Not done: these are internal types, not very important, and we do not modify them here.

File: db/src/query.ts (hunk 4 above)
Comment (arielshaqed) [reply db/src/query.ts]: Done (throughout...).

**Comment by arielshaqed** [comment]
@bergundy 
[I can haz PR approval?](https://imgflip.com/i/35xvxz)

File: db/src/db.ts [hunk 7]
```diff
@@ -8,6 +8,7 @@ import { Mutex } from 'async-mutex';
 import { ValueError } from './errors';
 import * as Q from './query';
 import { withTimeout, deferred, hrnano } from './utils';
+import { db as dbi } from '@binaris/shift-interfaces';
```
Comment (bergundy) [inline db/src/db.ts]: Please use the `Q` namespace instead.

Comment (bergundy) [reply db/src/db.ts]: (revert this file)

File: db/package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (bergundy) [reply db/package.json]: SGTM

File: db/src/query.ts [hunk 8]
```diff
@@ -347,50 +308,45 @@ export const value = Path.proxied(['value']);
...
+function checkFilters(...filters: any[]): Filter[] {
   if (filters.length === 0) {
     throw new IllegalArgumentError('Expected at least 1 filter');
   }
-  for (const f of filters) {
-    if (!f[filterSymbol]) {
-      throw new TypeError('Given filter is invalid');
-    }
-  }
+  filters.forEach(checkFilter);
```
Comment (bergundy) [inline db/src/query.ts]: Why did you change this?

File: db/src/db.ts (hunk 7 above)
Comment (arielshaqed) [reply db/src/db.ts]: Done.

File: db/src/query.ts (hunk 8 above)
Comment (arielshaqed) [reply db/src/query.ts]: Reverted.
